Test fix

jdistanislao opened the pull request:
**Tests**
Refactor tests using `testify` _Suite_ feature
Rewrite table-driven tests according to subtests guidelines (https://blog.golang.org/subtests)

**Bug fix**
Fix bug in `api/memegen.go:getImageFromRequest()`.
It uses the `mux.Vars()` method but `mux` is not correctly initialized and configured in the test. The initialization and configuration are spread across is the `api/controller.go` and `cmd/serve.go` files. Therefore it is very harmful to recreate a completely new `mux.NewRouter()` in the test without all the handling function used in those files. I also prefer to leave the old code commented on for further decisions. My suggestion is to refactor the routing and route handling in order o have a more testable and logically separated code.

jdistanislao committed:
add SUT, remove setup/tearDown calls from each test and use TestMain() func to handle the test lifecycle

jdistanislao committed:
refactoring api/memegen_test.go using testify suite feature

jdistanislao committed:
refactoring subtests in api/memegen_test.go

jdistanislao committed:
refactor TestMemeGenerate() file generation test using just the MemeFromRequest() response. Remove unnecessary dir creation in SetupSuite()

jdistanislao committed:
fix TestListGifs() and minor changes

jdistanislao committed:
use http.Request.FormValue() instead of mux.Vars() in order to avoid test failure beacause of no  mux initialization

jdistanislao committed:
refactor to testify test suite

jdistanislao committed:
split and refactoring tests in template_test.go in order to avoid unused variables

lavagetto commented on api/memegen_test.go:
```diff
@@ -13,40 +13,46 @@ import (
+func TestMain(m *testing.M) {
+	setUp()
+	code := m.Run()
+	tearDown()
```
given m.Run() can spawn goroutines, or worse panic/fail, it's a good idea to add a  defer tearDown() just below setUp() instead than here.

Your code will also not need the assignment of the code variable anymore.

lavagetto commented on api/memegen_test.go:
```diff
@@ -10,51 +10,66 @@ import (
 )
 
-var sut *MemeHandler
+const baseUrl string = "url"
```
Personal preference: while this is more _clean_, or better, more _pure_, I prefer in this case to have the values where the structure is declared, given we're not reusing them anywhere else. It makes the code easier to read in a single pass.

lavagetto commented on api/memegen_test.go:
```diff
@@ -85,17 +81,19 @@ func (s *MemeGenTestSuite) TestListGifs() {
+		s.Run(testName, func() {
 			if tc.Path != "" {
 				s.Sut.ImgPath = tc.Path
-			} else {
```
Removing this else will make it so that if we add more tests, we won't restore s.Sut.ImgPath to the original value we want to use as default.

The code could be nicer than it is in this version, probably, but the logic needs to be kept.

lavagetto commented on api/memegen_test.go:
```diff
@@ -126,23 +123,21 @@ func (s *MemeGenTestSuite) TestMemeGenerate() {
-				uid, err := s.Sut.UID(req)
-				s.Nil(err)
-				filePath := path.Join(s.Sut.OutputPath, fmt.Sprintf("%s.gif", uid))
+				locationPrefix := "/" + baseMemeUrl + "/"
```
nitpick: use fmt.Sprintf

non-nitpick: We should first test that response.Header contains a non-empty "Location" key.

lavagetto replied:
```suggestion
				locationPrefix := fmt.Sprintf("/%s/", s.Sut.MemeUrl
				locationHeader, ok := response.Header["Location"]
				s.True(ok, "response should include a Location header")
				s.NotEmpty(locationHeader, "response should include a Location header")
				// Extract location on disk from the Location Header
				fileName := locationHeader[0][len(locationPrefix):]
				
				
```

lavagetto commented on api/memegen_test.go:
```diff
@@ -82,17 +82,17 @@ func (s *MemeGenTestSuite) TestListGifs() {
 	for _, tc := range testCases {
 		testName := fmt.Sprintf("Path: %s - ContentType: %s - Status: %s - Body: %s", tc.Path, tc.ContentType, tc.Status, tc.Body)
 		s.Run(testName, func() {
+			s.Sut.ImgPath = baseImgPath
```
Ok I see my comment from the previous patch was taken into account here :+1:

lavagetto commented on img/template_test.go:
```diff
@@ -1,37 +1,30 @@
-	{"fixtures/earth.gif", false},
-	{"fixtures/non-existent.gif", true},
-	{"fixtures/badfile.gif", true},
+type TemplateTestSuite struct {
```
This is basically the same structure we have in ImageTestSuite. Adding a todo to unify the base struct in a comment might be a good idea.

jdistanislao replied:
Perfectly agree with you. I deliberately duplicate them as the first refactoring step, waiting for a clearer strategy.
Comment added.

lavagetto reviewed (changes requested):
This is overall very good, thanks for your work! I left a few comments on the individual patches, but most importantly, please improve the commit messages:
- Try to follow the 50/72 rule (first line of the commit under 50 chars, the rest under 72)
- Be a bit more descriptive

jdistanislao committed:
Minor refactor and add test to api/memegen_test.go

Better formatting for locationPrefix variable and add test for Location header before using it

Co-authored-by: Giuseppe Lavagetto <lavagetto@users.noreply.github.com>

jdistanislao committed:
Fix previous commit due a bad merge

jdistanislao committed:
add a TODO comment in img/template_test.go
further refactoring might be needed

lavagetto commented:
There are a few of my comments that still need to be taken into account. As for the git commits, for now I'll just reword them manually myself.
